### DISCUSSION TITLE
4/17 (월) 성하 - @ResponseBody VS ResponseEntity<T>

### DISCUSSION THREAD

**Discussion by sh111-coder** [woowacourse-study/2023-presentation-study#7] [Q&A]
Q&A

**Comment by jyeost** [comment]
1. 빌더를 사용할때 원하 Http 코드가 팩토리 메서드에 없으면 어떻게 사용해야 하나요?
2. @ResponseStatus() 에서 설정한 코드가 ResponseEntity.stauts() 의 코드와 다를때 어떤 결과가 나오나요?

**Reply by sh111-coder** [reply]
> 1. 빌더를 사용할때 원하는 Http 코드가 팩토리 메서드에 없으면 어떻게 사용해야 하나요?

ResponseEntity 클래스 내부를 살펴보니 바로 체이닝할 수 있는 HTTP 상태 코드는 다음과 같았습니다!
```java
1. ok() (200)
2. badReqeust() (400)
3. internalServerError (500)
4. accepted() (202)
5. created() (201)
6. noContent() (204)
7. notFound() (404)
8. unprocessableEntity() (422)
```
자주 사용하는 HTTP 상태 코드들은 직접 등록해놓은 것 같아요!

위에 해당하지 않는 HTTP 상태 코드를 사용하고 싶다면 `status(HttpStatus status)`를 사용해서
원하는 HTTP 상태 코드를 파라미터로 넣어주면 될 것 같아요!

ex : BAD_GATEWAY 사용하고 싶을 때
```java
return ResponseEntity.status(HttpStatus.HttpStatus.BAD_GATEWAY)
```

---
> 2. @ResponseStatus() 에서 설정한 코드가 ResponseEntity.stauts() 의 코드와 다를때 어떤 결과가 나오나요?

`ResponseEntity`가 상태 코드, 헤더 및 본문을 포함한 전체 HTTP 응답을 나타내고, 
@ResponseStatus는 단순 상태 코드만 나타내기 때문에 ResponsEntity가 @ResponseStatus보다 우선순위가 높게 적용되는 것 같아요!

제 능력 부족으로 원리 같은건 못 찾게 돼서 
`@ResponseStatus`보다 `ResponseEntity`가 우선순위가 높다! 는 것만 알면 될 것 같습니다.

**Comment by yeonkkk** [comment]
ResponseEntity 를 생성하는 방법으로 1. 생성자 2. 빌더 패턴 를 알려주셨어요.
그런데 빌더 패턴과 생성자를 모두 사용하는 이유가 궁금합니다. 
예를 들어 정적 팩토리 메서드를 사용하면, 생성자는 private으로 접근 제어자를 설정하여 생성자로 생성하는 것을 막는데, 이 경우 빌더 패턴과 생성자를 모두 사용 가능하도록 하는 것이 궁금해요.
혹시 생성자로 ResponseEntity를 꼭 생성해야하는 케이스가 있는지 궁금합니다.

**Reply by sh111-coder** [reply]
> 빌더 패턴과 생성자를 모두 사용하는 이유가 궁금합니다.

생성자의 파라미터가 빌더 체이닝 요소에 다 포함되어 있기 때문에
꼭 생성자로 ResponseEntity를 사용해야 하는 경우는 없는 것 같아요.

또 생성자를 막는 부분은 예로 들어주신 정적 팩토리 메소드와 빌더 패턴은 다르기 때문에
빌더 패턴을 사용한다고 꼭 생성자를 막아줘야하는 것은 아닌 것 같습니다!

'생성자와 빌더'를 같이 제공하는 이유는 
공식적인 것은 아니고, 나름대로 찾아봤는데
`다양한 상황에서의 사용자 편의성 고려` 측면이 있을 것 같아요!

매개변수가 적은 경우에는 생성자를 사용하여 간단하게 객체를 생성하고, 
매개변수가 많은 경우에는 빌더를 사용하여 더 유연하게 객체를 생성할 수 있을 것 같아요!

하지만 생성자를 사용할 경우에 하드 코딩하는 위험이 있을 수 있기 때문에 빌더를 권장한다고 합니다!
